src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/VisitService.kt

```diff
@@ -0,0 +1,169 @@
+            OffenderVisitBalanceAdjustment(
+              offenderBooking = offenderBooking,
+              adjustDate = visitDto.issueDate,
+              adjustReasonCode = VisitOrderAdjustmentReason.PVO_ISSUE,
+              remainingPrivilegedVisitOrders = -1,
+              previousRemainingPrivilegedVisitOrders = offenderVisitBalance.remainingPrivilegedVisitOrders,
+              commentText = "Created by PVB3 for an on-line visit booking",
+              //  authorisedStaffId = visitDto.staffId,
+              //  endorsedStaffId = visitDto.staffId,
+            )
+          )
+        }
+      } else {
+        if (offenderVisitBalance.remainingVisitOrders!! > 0) {
+          offenderVisitBalanceAdjustmentRepository.save(
+            OffenderVisitBalanceAdjustment(
+              offenderBooking = offenderBooking,
+              adjustDate = visitDto.issueDate,
+              adjustReasonCode = VisitOrderAdjustmentReason.VISIT_ORDER_ISSUE,
+              remainingVisitOrders = -1,
+              previousRemainingVisitOrders = offenderVisitBalance.remainingVisitOrders,
+              commentText = "Created by PVB3 for an on-line visit booking",
+              //  authorisedStaffId = visitDto.staffId,
+              //  endorsedStaffId = visitDto.staffId,
+              // TODO: If necessary, could follow stored proc logic and use staffId from:
+              //   select staff_id from staff_user_accounts where username = 'OMS_OWNER';
+            )
+          )
+        }
+      }
+    }
+  }
+
+  private fun mapVisitModel(visitDto: CreateVisitRequest, offenderBooking: OffenderBooking): Visit {
+
+    val visitType = visitTypeRepository.findById(VisitType.pk(visitDto.visitType)).orElseThrow()
+
+    return Visit(
+      offenderBooking = offenderBooking,
+      visitDate = LocalDate.from(visitDto.startTime),
+      startTime = visitDto.startTime,
+      endTime = LocalDateTime.of(LocalDate.from(visitDto.startTime), visitDto.endTime),
+      visitType = visitType,
+      visitStatus = visitStatusRepository.findById(VisitStatus.pk("SCH")).orElseThrow(),
+      location = agencyRepository.findById(visitDto.prisonId).orElseThrow(),
+      // TODO not yet sure if anything else is needed
+      // searchLevel = searchRepository.findById(SearchLevel.pk("FULL")).orElseThrow(),
+      // agencyInternalLocation = agencyInternalRepository.findById(-3L).orElseThrow(),
+      // commentText = "comment text",
+      // visitorConcernText = "visitor concerns",
+    )
+  }
+
+  private fun addVisitors(
+    visit: Visit,
+    offenderBooking: OffenderBooking?,
+    visitDto: CreateVisitRequest
+  ) {
+    val scheduledEventStatus = eventStatusRepository.findById(SCHEDULED_APPROVED).orElseThrow()
+
+    //  Add dummy visitor row for the offender_booking as is required by the P-Nomis view
+
+    visitVisitorRepository.save(
+      VisitVisitor(
+        visitId = visit.id!!,
+        offenderBooking = offenderBooking,
```

Comment:
maybe this should be mapped to `visit` rather than `visitId` - again feels weird you need to `visit.id!!`

Reply:
done